core/app/models/spree/variant.rb

```diff
@@ -73,7 +73,7 @@ class Variant < Spree::Base
 
     before_validation :set_cost_currency
     before_validation :set_price
-    before_validation :build_vat_prices, if: -> { rebuild_vat_prices? || new_record? }
+    before_validation :build_vat_prices, if: -> { rebuild_vat_prices? || new_record? && product }
 
     validate :check_price
```

Comment:
What about adding a `presence` validation on the `product` here? There are too many places where Solidus gives for granted that a product should be associated with a variant and I think it could make sense to don't allow saving a variant without a product associated. Also, this way users will have the right validation message attached to the invalid object.

Reply:
:+1:

Reply:
Hm, after reviewing the code I'm having second thoughts. How can we validate product presence in variant if product requires a master variant to be present?

The funny thing is that product automatically builds master variant, but I was unable to find `build_master` method definition. Maybe I'm missing something :frowning_face: 

https://github.com/solidusio/solidus/blob/adc75347b93fa9a51195051530b892f3d3379486/core/app/models/spree/product.rb#L74-L76

Any idea?

Reply:
`build_master` is a [method coming from has_one association](https://guides.rubyonrails.org/association_basics.html#methods-added-by-has-one). That variant will be associated with the product automatically so I think it will be valid, even with the proposed product presence validation.

Reply:
I see. :+1:

Reply:
Done